metadata/src/lib.rs

```diff
@@ -0,0 +1,529 @@
+        Self {
+            metadata_hash: [0; 32],
+            pallet_hashes: HashMap::new(),
+        }
+    }
+}
+
+impl Default for MetadataHashDetails {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        get_metadata_hash,
+        get_pallet_hash,
+    };
+    use frame_metadata::{
+        ExtrinsicMetadata,
+        PalletCallMetadata,
+        PalletMetadata,
+        RuntimeMetadataLastVersion,
+    };
+    use scale_info::meta_type;
+
+    // Define recursive types.
+    #[allow(dead_code)]
+    #[derive(scale_info::TypeInfo)]
+    struct A {
+        pub b: Box<B>,
+    }
+    #[allow(dead_code)]
+    #[derive(scale_info::TypeInfo)]
+    struct B {
+        pub a: Box<A>,
+    }
+
+    fn build_default_extrinsic() -> ExtrinsicMetadata {
+        ExtrinsicMetadata {
+            ty: meta_type::<()>(),
+            version: 0,
+            signed_extensions: vec![],
+        }
+    }
+
+    fn build_default_pallet() -> PalletMetadata {
+        PalletMetadata {
+            name: "Test",
+            storage: None,
+            calls: None,
+            event: None,
+            constants: vec![],
+            error: None,
+            index: 0,
+        }
+    }
+
+    fn build_default_pallets() -> Vec<PalletMetadata> {
+        let mut pallet_first = build_default_pallet();
+        pallet_first.name = "First";
+        pallet_first.calls = Some(PalletCallMetadata {
+            ty: meta_type::<u8>(),
+        });
+        let mut pallet_second = build_default_pallet();
+        pallet_second.name = "Second";
+        pallet_second.calls = Some(PalletCallMetadata {
+            ty: meta_type::<u16>(),
+        });
+        pallet_second.index = 1;
+
+        vec![pallet_first, pallet_second]
+    }
+
+    fn pallets_to_metadata(pallets: Vec<PalletMetadata>) -> RuntimeMetadataLastVersion {
+        RuntimeMetadataLastVersion::new(
+            pallets,
+            build_default_extrinsic(),
+            meta_type::<()>(),
+        )
+    }
+
+    #[test]
+    fn different_pallet_index() {
+        let pallets = build_default_pallets();
+        let mut pallets_swap = pallets.clone();
+
+        let metadata = pallets_to_metadata(pallets);
+
+        // Change the order in which pallets are registered.
+        pallets_swap.swap(0, 1);
+        pallets_swap[0].index = 0;
+        pallets_swap[1].index = 1;
+        let metadata_swap = pallets_to_metadata(pallets_swap);
+
+        let hash = get_metadata_hash(&metadata);
+        let hash_swap = get_metadata_hash(&metadata_swap);
+
+        // Changing pallet order must still result in a deterministic unique hash.
+        assert_eq!(hash.metadata_hash, hash_swap.metadata_hash);
+        assert_eq!(hash.pallet_hashes, hash_swap.pallet_hashes);
+
+        assert_eq!(
+            hex::encode(hash.metadata_hash),
+            "59f7633d9bf2daefaf6cd5788a5da8f7f787047e2d2d2ace4b3b142b7208f68f"
+        );
```

Comment:
I'm personally not fussed about validating that an actual hash is correct, and think we should just verify that whatever the hash is, it is consistent when pallets are moved around (ie, keep the test fairly focused on that one aspect).

(I can see us tweaking the hashing going forwards, and it would be a pain updating all of these hashes as we tweak and update the algorithm, anyway).